File: Python/Euler04/Euler04.py
# A palindromic number reads the same both
# ways. The largest palindrome made from
# the product of two 2-digit numbers is
# 9009 = 91 × 99.
# Find the largest palindrome made from
# the product of two 3-digit numbers.
# (Answer = 906609)

# Converts int to string and uses slicing to compare
# ~ 9,000,000 nanoseconds

import time

start = time.perf_counter_ns()

answer = 0

for i in range(1000, 900, -1):
    for j in range(1000, 900, -1):
        product = i * j
        textual = str(product)
        if textual == textual[::-1] and product > answer: # [::-1] returns reverse string
            answer = product
            val1 = i
            val2 = j

duration = time.perf_counter_ns() - start

print(answer)
print("Product of ", val1, " and ", val2)
print("Took: %s nanoseconds" % (duration))

File: C_Sharp/Euler04/Euler04.cs
/*A palindromic number reads the same both ways. 
The largest palindrome made from the product of 
two 2-digit numbers is 9009 = 91 × 99.
Find the largest palindrome made from the product 
of two 3-digit numbers. 
(Answer = 906609) */

// Uses math to reverse positive integers
// ~ 3500 ticks

using System;
using System.Diagnostics;

namespace Euler04name
{
    class Euler04
    {
        static int revnum(int num) // reverse number mathmatically (only positive integers)
        {
            int reverse = 0;
            int endDigit = 0;
            while (num > 0)
            {
                endDigit = num % 10;
                reverse = (reverse * 10) + endDigit;
                num = num / 10;  // remainder is rounded
            }
            return reverse;
        }

        static void Main(string[] args)
        {
            var watch = new System.Diagnostics.Stopwatch();
            watch.Start();

            int temp = 0;
            int temp2 = 0;
            int answer = 0;
            int val1 = 0;
            int val2 = 0;

            for (int i = 1000; i > 900; i--)
            {
                for (int j = 1000; j > 900; j--)
                {
                    temp = i * j;
                    temp2 = revnum(temp);

                    if (temp == temp2 && temp > answer)
                    {
                        answer = temp;
                        val1 = i; // store values multiplied
                        val2 = j;
                    }
                }
            }
            watch.Stop();
            Console.WriteLine("Answer: {0}", answer);
            Console.WriteLine("Product of {0} & {1}", val1, val2);
            Console.WriteLine("Took: {0} ticks", watch.ElapsedTicks);
            Console.WriteLine("Took: {0} ms", watch.ElapsedMilliseconds);
        }
    }
}
